text/0013-portable-pdb.md

```diff
@@ -0,0 +1,105 @@
+
+# Summary
+
+This RFC proposes new protocol fields to help with symbolicating .NET stack traces that use Portable PDBs internally for
+symbolication.
+
+# Motivation
+
+To symbolicate a stack frame using a Portable PDB, we need the following data:
+
+- IL Offset
+- Method Index
+
+Additionally, fetching a Portable PDB from an external Symbol server following the [SSQP specification](https://github.com/dotnet/symstore/blob/main/docs/specs/SSQP_Key_Conventions.md#portable-pdb-signature) needs to have:
+
+- PDB File name
+- PDB Id
+- PDB Checksum
+
+Currently the `SentryStackFrame` type of the `sentry-dotnet` SDK has an
+[`instruction_offset` field](https://github.com/getsentry/sentry-dotnet/blob/57044ff52320c82cf1ca22cceee7125dfb9d1423/src/Sentry/SentryStackFrame.cs#L128-L135),
+which is itself not yet documented on [`develop.sentry.dev`](https://develop.sentry.dev/sdk/event-payloads/stacktrace/#frame-attributes).
+
+## NuGet Symbol Server
+
+The NuGet symbol server (`https://symbols.nuget.org/download/symbols/{query}`) requires the PDB checksum to be provided via a `SymbolChecksum` header.
+If that header is not provided, the server will always respond with a `403` error code. Otherwise, it uses a `404` code when the symbol is not found.
+
+Examples:
+
+```
+> curl https://symbols.nuget.org/download/symbols/microsoft.maui.pdb/10f7f174e11949f587c4d0b617742d31FFFFFFFF/microsoft.maui.pdb -i
+HTTP/1.1 403 Forbidden
+
+> curl https://symbols.nuget.org/download/symbols/microsoft.maui.pdb/10f7f174e11949f587c4d0b617742d31FFFFFFFF/microsoft.maui.pdb -H "SymbolChecksum: SHA256:74f1f71019e1f5197c4d0b617742d31b515f041c4f62a41bfc25ebb05a7fbff3" -i
+HTTP/1.1 404 Not Found
+```
+
+The details of this header do not seem to be publicly documented, but can be traced by looking through the `symstore` code:
+
+- PdbChecksum: https://github.com/dotnet/symstore/blob/aa44862e5028cb7595bbd474da1d63e8c24bf718/src/Microsoft.FileFormats/PE/PEStructures.cs#L375
+- PE Key: https://github.com/dotnet/symstore/blob/aa44862e5028cb7595bbd474da1d63e8c24bf718/src/Microsoft.SymbolStore/KeyGenerators/PEFileKeyGenerator.cs#L75
+- Portable PDB Key: https://github.com/dotnet/symstore/blob/aa44862e5028cb7595bbd474da1d63e8c24bf718/src/Microsoft.SymbolStore/KeyGenerators/PortablePDBFileKeyGenerator.cs#L85
+- Key Format: https://github.com/dotnet/symstore/blob/aa44862e5028cb7595bbd474da1d63e8c24bf718/src/Microsoft.SymbolStore/KeyGenerators/KeyGenerator.cs#L166-L177
+- `SymbolChecksum` Header: https://github.com/dotnet/symstore/blob/8a7c47ac74302510f839cc2361ca591b6f4df542/src/Microsoft.SymbolStore/SymbolStores/HttpSymbolStore.cs#L112
+
+# Options Considered
+
+## Stack Trace Interface Extensions
+
+We propose to add the following to the [Stack Trace Interface Frame Attributes](https://develop.sentry.dev/sdk/event-payloads/stacktrace/#frame-attributes):
+
+- `instruction_offset: Number`, as it is currently used by the `sentry-dotnet` SDK.
```

Comment:
`instruction_offset` is a good generic name for such a field in principle. Given that the old fields for this have been long removed and were inactive, there is no risk of collisions.

If we want to be more specific, we could also name this something like `function_offset`, which would further differentiate the semantics of this field to WASM's addressing mode.

Reply:
How about this:
 - there is a field `instruction_addr`.
 - there is a field `addr_mode` which may be `abs` or `rel:n`.
 - there is an optional field `function_index` (name subject to change).

 `instruction_addr` is then interpreted as follows:

 - `addr_mode = abs`: `instruction_addr` is an absolute address
 - `addr_mode = rel:n` and `function_index` is missing: `instruction_addr` is relative to module `n` (the WASM case)
 - `addr_mode = rel:n` and `function_index = k`: `instruction_addr` is relative to function `k` in module `n` (in other words, an IL offset)

Reply:
I don't hate this but I wonder if we then don't want to have a `funcrel:n:m` where `n` is the module index and `m` is the function index?

Reply:
We were also talking about that last week, and did not really like it because it was too "stringly typed". And parsing things from a specially formatted string is strictly worse than just having proper JSON properties for it.

However I do recognize that the whole problem boils down to having a properly discriminated union:
```rust
enum Addr {
  Absolute(u64),
  ModuleRelative { module_idx: u32, addr: u64 },
  FunctionRelative { module_idx: u32, function_idx: u32, addr: u64 },
}
```